### PR TITLE
V2: Freeze parts of model (Transfer learning)

### LINKED ISSUE #651: [TODO]: Add way to freeze parts of the model during training.
**Notes**
The v1 README has some details about how `--checkpoint_frzn`, `--frzn_ffn_layers`, and `--freeze_first_only` worked. 

Primarily this functionality is [in](https://github.com/chemprop/chemprop/blob/d58d877fd1d652929b0216e7b0cc11355664fd97/chemprop/utils.py#L195) `chemprop.utils.load_frzn_model`. But also when the model is [created](https://github.com/chemprop/chemprop/blob/d58d877fd1d652929b0216e7b0cc11355664fd97/chemprop/models/model.py#L86), the params that are frozen need `.requires_grad = False`.

### PR DESCRIPTION
**Pull request by KnathanM**
### Motivation

Transfer learning is sometimes useful when there isn't a lot of data for target a user wants to predict. In this case, the user can train a Chemprop model for a target that does have sufficient data, and then transfer the weights of the message passing layer (and optionally for some layers of the FFNN) to a new model, freeze them and use their limited data to train a smaller number of parameters. This will work well if the learned representation output by the message passing has meaningful descriptors on the input molecule. 

### Methodology

Loading in weights from a pre-trained model into a current model is simply done using `model.load_state_dict(state_dict)`. Here I opted to load in the entire pre-trained model as the starting point using `MPNN.load_from_file()`. This has two potential shortcomings: 
(1) The weights that should be trained (e.g. the unfrozen layers of the FFNN) won't be randomly initialized as they are read in from the previous training. I believe in v1, a new model was first created with random weights and then only the weights to be frozen were read in. 
(2) The user can't adjust any part of the architecture. This would be a problem if they wanted to adjust how many hidden layers in the FFNN for example. Perhaps we call this as "advanced" and expect users to do the transfer learning manually in a notebook if they want this functionality. If we didn't load in the whole pre-trained model, it would give more flexibility, but also require more code. 

Once the weights are loaded, they are "frozen" by telling torch to not calculate gradients for them via `.requires_grad_(False)`. I have seen discussion online about whether these frozen parameters should be excluded from being passed to the optimizer by using `opt = optim.Adam(filter(lambda p: p.requires_grad, self.parameters()), self.init_lr)` instead of `opt = optim.Adam(self.parameters(), self.init_lr)`, but opinions were mixed and my tests showed this wasn't necessary. 

After each `Linear` layer in both the message_passing and predictor submodules, a dropout is applied during training. This should be turned off for frozen layers as this random zeroing of weights produces non-deterministic outputs of the frozen layer. Dropout is automatically turned off when the whole model is set to evaluation model (`model.eval()`) but we can also turn it off manually by using `.eval()` for the specific dropout layers we want to turn off. A difficulty here is that before the lightning Trainer calls `model.training_step()` in `trainer.fit()` it will set the whole model to train. This means the dropout layers have to be set to eval within the model itself. This applies whenever the lightning trainer is used, both in the CLI and in notebook use. I chose to do this in `model.fingerprint()` because both `model.forward()` and `model.training_step()` call this function. 

### Tests

If this approach looks good, I will add tests and potentially an example notebook to this PR.

### Questions

* Is the learned representation of a molecule the vector before batch normalization or after?
Currently there is a `nn.BatchNorm1d` layer that gets applied to the output of the aggregation layer before it is returned in `fingerprint()`. If we want the learned representation of a molecule to be fixed by freezing, then the batch normalization needs to be frozen as well. 
But the intent of the batch normalization is to stabilize the input to the FFNN to help the FFNN learn faster. Perhaps we should move move the batch normalization into the predictor or at least out of `fingerprint()`. 
On a related note, in the case of multicomponent input (e.g. solute + solvent), v1 had the option to only freeze the weights for one of them. I plan to add this later (maybe v2.1). If batch norm goes in `fingerprint()` and we have only frozen one of the message_passing modules, we can't really freeze the batch norm. So the learned representation of the component with the frozen message_passing will change due to the normalization with the other component's learned representation, unless the batch normalization is moved out of `fingerprint()`. 
* I'm worried that adding `if self.message_passing.freeze` to every `fingerprint()` call will slow down training unnecessarily when not doing transfer learning. Is there a way to avoid this?
* Are the two concerns I raise about using `model.load_state_dict(state_dict)` enough reason to overwrite the model weights manually?

Closes #651

## Checklist
- [x] linted with flake8?
- [ ] unit tests added?

### CODE REVIEW

File: chemprop/nn/predictors.py [hunk 1]
```diff
@@ -80,6 +80,7 @@ def __init__(
             input_dim, n_tasks * self.n_targets, hidden_dim, n_layers, dropout, activation
         )
         self.criterion = criterion or self._default_criterion
+        self.freeze = 0
```
Comment (hwpang) [inline chemprop/nn/predictors.py:83]: This is a nitpick, but could we change this to something to reflect that this is a number as `self.freeze` is bool in other locations? Something like `self.freeze_layers`?

**Comment by davidegraff** [comment]
Thanks for this write-up! You've raised some valid concerns about the difficulty of reimplementing the v1 frozen behavior in v2. Here are some of my thoughts:

> Once the weights are loaded, they are "frozen" by telling torch to not calculate gradients for them via `Parameter.requires_grad_(False)`. I have seen discussion online about whether these frozen parameters should be excluded from being passed to the optimizer by using
> ```
> opt = optim.Adam(filter(lambda p: p.requires_grad, self.parameters()), self.init_lr)
> ```
> instead of
> ```
> opt = optim.Adam(self.parameters(), self.init_lr)
> ```
> but opinions were mixed and my tests showed this wasn't necessary.

Per [this post](https://discuss.pytorch.org/t/passing-to-the-optimizers-frozen-parameters/83358/2), any frozen parameters (i.e., `requires_grad = False`) will be filtered during the optimizer step (see [here](https://github.com/pytorch/pytorch/blob/b7b99ab0c8f82100177729b9751481852d83e77e/torch/optim/sgd.py#L95-L96)). So filtering them during optimizer initialization is unnecessary.

> After each Linear layer in both the `message_passing` and `predictor` submodules, a dropout layer is applied. This should be turned off for frozen layers as this random zeroing of weights produces non-deterministic outputs of the frozen layer. Dropout is automatically turned off when the whole model is set to evaluation model (`eval()`) but we can also turn it off manually by using `eval()` for the specific dropout layers we want to turn off. A difficulty here is that before the lightning Trainer calls `MPNN.training_step()` in `trainer.fit()` it will set the whole model to train. This means the dropout layers have to be set to eval within the model itself. This applies whenever the lightning trainer is used, both in the CLI and in notebook use. I chose to do this in `MPNN.fingerprint()` because both `MPNN.forward()` and `MPNN.training_step()` call this function.
>
> I'm worried that adding if `self.message_passing.freeze` to every `MPNN.fingerprint()` call will slow down training unnecessarily when not doing transfer learning. Is there a way to avoid this?

This is a good point, but I don't think this is the right execution. I'm not clear on why we need to turn off gradients _at each step_ rather than prior to model training? I.e., why can't we do this:
```python
model = load_model(...)
if args.freeze_encoder:
    model.message_passing.apply(lambda module: module.requires_grad_(False))
if args.freeze_ffn:
    model.predictor[:-1].apply(lambda module: module.requires_grad_(False))
...
trainer.fit(model)
```
re: turning off dropout for frozen modules, that's something I didn't think about. But turning _off_ dropout is functionally equivalent to setting the probability to 0. That is:
```python
net: nn.Module = ...
net.apply(lambda m: setattr(m, 'p', 0.) if isinstance(m, nn.Dropout) else None)
```
is equivalent to (assuming no batch norm, etc.):
```python
net: nn.Module = ...
net.eval()
```
during evaluation. The difference is that the first works during training as well.

> - Is the learned representation of a molecule the vector before batch normalization or after?
Currently there is a `nn.BatchNorm1d` layer that gets applied to the output of the aggregation layer before it is returned in `fingerprint()`. If we want the learned representation of a molecule to be fixed by freezing, then the batch normalization needs to be frozen as well.
> But the intent of the batch normalization is to stabilize the input to the FFNN to help the FFNN learn faster. Perhaps we should move move the batch normalization into the predictor or at least out of `fingerprint()`.
On a related note, in the case of multicomponent input (e.g. solute + solvent), v1 had the option to only freeze the weights for one of them. I plan to add this later (maybe v2.1). If batch norm goes in `fingerprint()` and we have only frozen one of the `message_passing` modules, we can't really freeze the batch norm. So the learned representation of the component with the frozen message_passing will change due to the normalization with the other component's learned representation, unless the batch normalization is moved out of `fingerprint()`.

There's no single definition of "learned representation" in the context of a deep neural network. Each intermediate vector can be considered a learned representation, and the ML/chemistry field has sort of adopted calling the aggregated output of a GNN a "learned fingerprint" due to its parallels to a traditional fingerprinting algorithm (to see this, consider the ECFP algorithm and compare it to a MPNN on molecular graphs. You can write the ECFP algorithm as a type of MPNN!) All of that said, there's no strict requirement for the batch norm being inside `fingerprint()`. If it makes this implementation cleaner, you can feel free to move it into `forward()` and `latent()`.

**Comment by KnathanM** [comment]
> This is a good point, but I don't think this is the right execution. I'm not clear on why we need to turn off gradients at each step rather than prior to model training?

From my tests I thought was necessary, but I tried again and your suggestion worked just fine. I think the dropout was changing my fingerprints between epochs which made me think I had to keep turning off gradients. That was when I was using dropout.eval(). By setting the dropout probability to 0, I have no issue. 

I still want to add a test for these commands to the PR, but will do that tomorrow.

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -90,20 +90,21 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         help="The number of batches between each logging of the training loss.",
     )
     parser.add_argument(
-        "--checkpoint-frzn",
-        help="Path to model checkpoint file to be loaded for overwriting and freezing weights.",
+        "--model-frzn",
+        type=Path,
+        help="Path to model file or checkpoint file to be loaded for overwriting and freezing weights. If given, the fingerprinting layer is frozen. Parts of the ffn layer can be frozen using '--frzn-ffn-layers'.",
```
Comment (hwpang) [inline chemprop/cli/train.py:95]:
```suggestion
        help="Path to model file (.pt) or checkpoint file (.ckpt) to be loaded for overwriting and freezing weights. If given, the fingerprinting layer is frozen. Parts of the ffn layer can be frozen using '--frzn-ffn-layers'.",
```

File: chemprop/cli/train.py [hunk 3]
```diff
@@ -90,20 +90,21 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
         "--frzn-ffn-layers",
         type=int,
         default=0,
-        help="Overwrites weights for the first n layers of the ffn from checkpoint model (specified checkpoint_frzn), where n is specified in the input. Automatically also freezes mpnn weights.",
-    )
-    parser.add_argument(
-        "--freeze-first-only",
-        action="store_true",
-        help="Determines whether or not to use checkpoint_frzn for just the first encoder. Default (False) is to use the checkpoint to freeze all encoders. (only relevant for number_of_molecules > 1, where checkpoint model has number_of_molecules = 1)",
+        help="Overwrites weights for the first n layers of the ffn from '--model_frzn', where n is specified in the input.",
```
Comment (hwpang) [inline chemprop/cli/train.py:101]:
```suggestion
        help="Overwrites weights for the first n layers of the ffn from '--model_frzn', where `n` is specified in the input.",
```

**Review by hwpang — CHANGES_REQUESTED** [review]
Can you add a test to test these? Maybe somewhere in `chemprop/tests/cli/test_cli_regression_mol.py` to test a simple transfer learning

**Comment by KnathanM** [comment]
The test I added checks if the weights of part of the message passing don't change during training if frozen. It also checks that part of the FFN isn't changed when frozen. I checked that if the freezing command line arguments aren't specified then 1 epoch is still enough to change those weights. 

In addition to preventing the weights from changing, it is also important that freezing the model turns off dropout. I was unable to add a test for that though because turning that off is internal to the training...

**Review by hwpang — APPROVED** [review]
LGTM!
